Sources/TestingInternals/Discovery.cpp

```diff
@@ -158,20 +163,89 @@ struct SWTTypeMetadataRecord {
   }
 };
 
+/// Specifies the address range corresponding to a section.
+struct SWTMetadataSectionRange {
+  uintptr_t start;
+  size_t length;
+};
+
+/// Identifies the address space ranges for the Swift metadata required by the
+/// Swift runtime.
+///
+/// - Note: Most of the fields in this structure are zeroed on Apple platforms
+///   since we don't need them. On Linux and Windows, they are generally
+///   populated (as provided by the Swift runtime.) The following fields can be
+///   relied upon:
+///
+///   - ``version``
+///   - ``baseAddress`` (unless `SWT_NO_DYNAMIC_LINKING` is defined.)
+///   - ``swift5_type_metadata``
+///   - ``swift5_tests`` (if ``version`` is at least equal to
+///     ``SWT_METADATA_SECTION_MINIMUM_VERSION_WITH_TESTS``.)
+struct SWTMetadataSections {
+  uintptr_t version;
+  std::atomic<const void *> baseAddress;
+
+  void *unused0;
+  void *unused1;
+
+  SWTMetadataSectionRange swift5_protocols;
+  SWTMetadataSectionRange swift5_protocol_conformances;
+  SWTMetadataSectionRange swift5_type_metadata;
+  SWTMetadataSectionRange swift5_typeref;
+  SWTMetadataSectionRange swift5_reflstr;
+  SWTMetadataSectionRange swift5_fieldmd;
+  SWTMetadataSectionRange swift5_assocty;
+  SWTMetadataSectionRange swift5_replace;
+  SWTMetadataSectionRange swift5_replac2;
+  SWTMetadataSectionRange swift5_builtin;
+  SWTMetadataSectionRange swift5_capture;
+  SWTMetadataSectionRange swift5_mpenum;
+  SWTMetadataSectionRange swift5_accessible_functions;
+  SWTMetadataSectionRange swift5_runtime_attributes;
+
+  // TODO: Add this field to the MetadataSections definition in the Swift runtime.
+  SWTMetadataSectionRange swift5_tests;
+};
+
+/// The minimum value of ``SWTMetadataSections/version`` if that instance of
+/// ``SWTMetadataSections`` contains the ``SWTMetadataSections/swift5_tests``
+/// field.
+static constexpr uintptr_t SWT_METADATA_SECTION_MINIMUM_VERSION_WITH_TESTS = 9999;
+
 #if defined(SWT_NO_DYNAMIC_LINKING)
 #pragma mark - Statically-linked implementation
 
 // This environment does not have a dynamic linker/loader. Therefore, there is
 // only one image (this one) with Swift code in it.
 // SEE: https://github.com/apple/swift/tree/main/stdlib/public/runtime/ImageInspectionStatic.cpp
 
-extern "C" const char sectionBegin __asm("section$start$__TEXT$__swift5_types");
-extern "C" const char sectionEnd __asm("section$end$__TEXT$__swift5_types");
+extern "C" const char typesSectionBegin __asm("section$start$__TEXT$__swift5_types");
+extern "C" const char typesSectionEnd __asm("section$end$__TEXT$__swift5_types");
+
+extern "C" const char testsSectionBegin __asm("section$start$__DATA_CONST$__swift5_tests");
+extern "C" const char testsSectionEnd __asm("section$end$__DATA_CONST$__swift5_tests");
+
+// Ensure the tests section is actually emitted.
+__attribute__((used)) __attribute__((section("__DATA_CONST,__swift5_tests")))
+constinit const SWTTestGetter ensureTestSectionIsEmitted = std::make_pair(nullptr, nullptr);
```

Comment:
This seems like it would be limited to MachO? Should this region also have a restriction? Or do you intend to re-declare these for ELF, COFF, and WASM?

Reply:
You can ignore this section; it's intentionally only functional with Mach-O. Linux and Windows should always use the code path below that uses `swift_enumerateAllMetadataSections()`.

Embedded Swift is an open question right now; we're only just starting to look at what's feasible. Presumably, solutions for Embedded Swift will look similar to this one, but will need to be arch- or platform-specific.